File: src/data/courses/dsa/linked-lists/two-pointer-pattern/cycle-start-node/code/python/cycle_start_node.py
# ============================================================
# Cycle Start Node
# ============================================================
class Solution:
    def detectCycle(self, head):
        slow = fast = head
        while fast and fast.next:
            slow, fast = slow.next, fast.next.next
            if slow == fast:
                ptr = head
                while ptr != slow: ptr, slow = ptr.next, slow.next
                return ptr
        return None

File: src/data/courses/dsa/linked-lists/two-pointer-pattern/cycle-start-node/code/cpp/cycle_start_node.cpp
// ============================================================
// Linked List Cycle Start Node
// ============================================================
struct ListNode { int val; ListNode* next; ListNode(int v):val(v),next(nullptr){} };

class Solution {
public:
    ListNode* detectCycle(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head;

        // Phase 1: Find meeting point
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
            if (slow == fast) {
                // Phase 2: Find cycle start
                ListNode* ptr = head;
                while (ptr != slow) {
                    ptr = ptr->next;
                    slow = slow->next;
                }
                return ptr; // Cycle start
            }
        }
        return nullptr; // No cycle
    }
};
